bundles/config/org.eclipse.smarthome.config.discovery.usbserial/src/main/java/org/eclipse/smarthome/config/discovery/usbserial/internal/UsbSerialDiscoveryService.java

```diff
@@ -0,0 +1,160 @@
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link DiscoveryService} for discovering USB devices with an associated serial port.
+ * <p/>
+ * This discovery service is intended to be used by bindings that support USB devices, but do not directly talk to the
+ * USB devices but rather use a serial port for the communication, where the serial port is provided by an operating
+ * system driver outside the scope of Eclipse SmartHome. Examples for such USB devices are USB dongles that provide
+ * access to wireless networks, like, e.g., Zigbeee or Zwave dongles.
+ * <p/>
+ * This discovery service provides functionality for discovering added and removed USB devices and the corresponding
+ * serial ports. The actual {@link DiscoveryResult}s are then provided by {@link UsbSerialDiscoveryParticipant}s, which
+ * are called by this discovery service whenever new devices are detected or devices are removed. Such
+ * {@link UsbSerialDiscoveryParticipant}s should be provided by bindings accessing USB devices via a serial port.
+ * <p/>
+ * This discovery service requires a component implementing the interface {@link UsbSerialDiscovery}, which performs the
+ * actual serial port and USB device discovery (as this discovery might differ depending on the operating system).
+ *
+ * @author Henning Sudbrock - initial contribution
+ */
+@NonNullByDefault
+@Component(immediate = true, service = { DiscoveryService.class,
+        UsbSerialDiscoveryService.class }, configurationPid = "discovery.usbserial")
+public class UsbSerialDiscoveryService extends AbstractDiscoveryService implements UsbSerialDiscoveryListener {
+
+    private final Logger logger = LoggerFactory.getLogger(UsbSerialDiscoveryService.class);
+
+    private final Set<UsbSerialDiscoveryParticipant> discoveryParticipants = new CopyOnWriteArraySet<>();
+
+    @Nullable
+    private UsbSerialDiscovery usbSerialDiscovery;
+
+    public UsbSerialDiscoveryService() {
+        super(5);
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addUsbSerialDiscoveryParticipant(UsbSerialDiscoveryParticipant participant) {
+        this.discoveryParticipants.add(participant);
+    }
+
+    protected void removeUsbSerialDiscoveryParticipant(UsbSerialDiscoveryParticipant participant) {
+        this.discoveryParticipants.remove(participant);
+    }
+
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC)
+    protected void setUsbSerialDiscovery(UsbSerialDiscovery usbSerialDiscovery) {
+        this.usbSerialDiscovery = usbSerialDiscovery;
+        usbSerialDiscovery.registerDiscoveryListener(this);
+        if (isBackgroundDiscoveryEnabled()) {
+            usbSerialDiscovery.startBackgroundScanning();
+        }
+    }
+
+    protected void unsetUsbSerialDiscovery(UsbSerialDiscovery usbSerialDiscovery) {
```

Comment:
Dynamic service reference is a bit tricky in OSGi since no strict ordering of execution is followed. If you have two implementations of `UsbSerialDiscovery` (let's say A and B), binding callbacks can be called before the unbinding callbacks. That is, the order could be set(A) set(B) unset(A) and when unset is executed the `this.usbSerialDiscovery` will be null even though  it is not expected since the field should only be null, if the injected instance (in this scenario, B) is deinjected.

It makes even situation worse, if multiple threads start invoking the callbacks together. Since OSGi as a container doesn't restrict any ordering for dynamic service references, the proper synchronization needs to be introduced by the consumer.

I personally use the following pattern (but not necessarily, you have to. You got the idea and you can use any suitable functionality that suffices your need. For example, `AtomicReference` is also a very good fit in this scenario.)

```java
private volatile IA a;
protected (synchronized) void setA(IA a) {
   ....
}
protected (synchronized) void unsetA(IA a) {
   if (this.a == a) {
       ....
       this.a = null;
   }
}
```

Reply:
Thanks @amitjoy for this hint. I like the pattern that you suggest, but I think that it needs to be augmented a little, because in the scenario that you describe the set(B) is an implicit unset(A), which has to be taken care of in the set method as well (i.e., the unregistration of A when setting B). I will change this and would be glad if you had a second look on it after my change.